src/libstd/ffi/c_str.rs

```diff
@@ -718,6 +719,81 @@ impl From<Box<CStr>> for CString {
     }
 }
 
+#[stable(feature = "cstring_from_vec_of_nonzerou8", since = "1.39.0")]
+impl From<Vec<NonZeroU8>> for CString {
+    /// Converts a [`Vec`]`<`[`NonZeroU8`]`>` into a [`CString`] without
+    /// copying nor checking for inner null bytes.
+    ///
+    /// [`CString`]: ../ffi/struct.CString.html
+    /// [`NonZeroU8`]: ../num/struct.NonZeroU8.html
+    /// [`Vec`]: ../vec/struct.Vec.html
+    #[inline]
+    fn from(v: Vec<NonZeroU8>) -> CString {
+        /// # Safety
+        ///
+        ///   - for any `length: usize`, it must be sound to transmute between
+        ///     `[T; length]` and `[U; length]`.
+        ///
+        ///   - for any `cap: usize`, `Layout<[T; cap]>` needs to be equal to
+        ///     `Layout<[U; cap]>` (for the allocator)
+        #[inline]
+        unsafe
+        fn transmute_vec<T, U>(v: Vec<T>) -> Vec<U> {
+            // necessary conditions for `Layout<[T; N]> == Layout<[U; N]>`
+            {
+                assert_eq!(
+                    mem::size_of::<T>(),
+                    mem::size_of::<U>(),
+                );
+                assert_eq!(
+                    mem::align_of::<T>(),
+                    mem::align_of::<U>(),
+                );
+            }
+            // # Safety
+            //
+            //   - this is the pattern blessed by transmute's documentation
+            //     (and replacing `mem::forget` by `ManuallyDrop`, _c.f._,
+            //     issue #62553)
+            //
+            //   - `transmute_vec` contract guarantees sound transmuting
+            //     between `[T; length]` and `[U; length]`
+            let length: usize = v.len();
+            let capacity: usize = v.capacity();
+            let ptr: *mut T = {
+                // end-of-scope `mem::forget` but without aliasing problems
+                let mut v = mem::ManuallyDrop::<Vec<T>>::new(v);
+                let v: &mut Vec<T> = &mut *v;
+                v.as_mut_ptr() // cannot be aliased
+            };
+            Vec::<U>::from_raw_parts(ptr as *mut U, length, capacity)
+        }
+
+        let v = unsafe {
+            // # Safety
+            //
+            // For all `N: usize`, transmuting between `[NonZeroU8; N]` and
+            // `[u8; N]` is sound since:
+            //
+            //   - `[NonZeroU8; N]` and `[u8; N]` have the same layout in
+            //     memory;
+            //
+            //   - any valid representation for a `NonZeroU8` is a valid
+            //     representation for a `u8` (loosened representation).
+            transmute_vec::<NonZeroU8, u8>(v)
+        };
+        unsafe {
+            // # Safety
+            //
+            //   - `v` cannot contain null bytes, given the type-level
+            //     invariant of `NonZeroU8` (this would still apply even if
+            //     this invariant was a safety invariant and not a validity
```

Comment:
Nit: strike the hypothetical. I am reasonably sure you do need the safety invariant here. Most operations on vectors don't assert validity of any elements except perhaps the ones they move into or out of the vec. Of course, a `Vec<T>` is only *safe* if elements `0..len` are safe at `T`, so the soundness argument can rest on that plus the safety/validity (same thing) of the `NonZeroU8`.

(Insert standard disclaimer about all of these details being not yet ratified here.)

Reply:
You're right, I will remove the part between brackets then (I'm still glad I put it, it is good to clearly think about these things before merging).